akiraux/Akira
# Initial Implementation for Pencil Tool (Freehand Drawing)

## Summary / How this PR fixes the problem?

The Pencil Tool will allow users to draw freehand shapes and squiggles.

When drawing such a canvas item, all the users pointer events would be captured and stored. When rendering these points, we join them using a polyline. While this makes the task of creating such squiggles easy, storing large amounts of points becomes a problem.

To reduce the number of points, we use **Schneider's Algorithm**. This algorithm will compress the set of points into a set of bezier curves that are very close to what the user originally drew.

Using this algorithm will do two things-
1. Reduce the number of points.
2. Make the drawing more aesthetically pleasing. (because bezier curves are very smooth)..

##### Reference
1. The Schneider's Algorithm can be found in the book Graphics Gems I. I found a pdf [here](https://www.pdfdrive.com/graphics-gems-1-e38625848.html)
2. The actual code was a little hard to develop. So I referred [this](https://github.com/soswow/fit-curve/blob/master/src/fit-curve.js) repo. The code here is in JavaScript which I translated to Vala as best as I could. I also ran this implementation and my own one on different inputs and checked their outputs. They matched more or less.

#### Summary of major changes I made

- Created **Lib.Modes.FreehandMode** to get user events for creating freehand drawings.
- Created **Models.FreehandModel** to store and process the raw points from user events.
- Used *Schneider's Algorithm** to optimize the number of points needed to represent the drawing. Approximates the raw points into a set of bezier curves.

## Steps to Test

- Press ```P``` or use the **Insert** menu to start the pencil tool.
- Hold down your mouse button and get crazy.
- As soon as you are done drawing, release the mouse button. The curve should have formed.
- Double clicking on this item will show the control points of bezier curves formed.

## Screenshots 

https://user-images.githubusercontent.com/64722310/146642939-5334d9b2-70ac-474d-a943-b20e1cfd6a87.mp4

## Known Issues / Things To Do
